src/main/java/edu/wpi/grip/core/Connection.java

```diff
@@ -31,11 +34,33 @@ public Connection(EventBus eventBus, OutputSocket<? extends T> outputSocket, Inp
         checkNotNull(outputSocket);
         checkNotNull(eventBus);
 
+        if (willCreateInfiniteLoop(outputSocket, inputSocket)) {
+            throw new InfiniteLoopException(outputSocket, inputSocket);
+        }
+
         inputSocket.setValue(outputSocket.getValue());
 
         eventBus.register(this);
     }
 
+    /**
+     * Determines if this connection will create an infinite loop
+     *
+     * @param outputSocket The input socket to search for sockets
+     * @param inputSocket  The output socket to search for sockets
+     * @return False if this will create an infinite loop
+     */
+    private boolean willCreateInfiniteLoop(OutputSocket<?> outputSocket, InputSocket<?> inputSocket) {
+        // There can't be any loops between two sockets that aren't owned by steps
+        if (!outputSocket.getStep().isPresent()) return false;
+        checkArgument(inputSocket.getStep().isPresent(), "An input socket must be owned by a step. Only output sockets can be owned by a Source");
```

Comment:
Wouldn't you just return false in this case?  If we ever have a situation where an inputsocket isn't a step (which we don't currently), there's no way it could be in a loop.  I'm not sure if this bit of extra validation should be in this method.